make kata-tarball silently omits crucial pieces

# Description of problem

Build kata artifacts with `make kata-tarball`

# Expected result

Builds all necessary artifacts to deploy kata,

# Actual result

Although the `make kata-tarball` completes without apparent error, the resulting `kata-static.tar.xz` is missing the rootfs image file:

```
$ tar tvfJ kata-static.tar.xz 
drwxr-xr-x dwg/dwg           0 2022-02-24 16:09 ./
drwxr-xr-x dwg/dwg           0 2022-02-24 16:09 ./opt/
drwxr-xr-x dwg/dwg           0 2022-02-24 16:09 ./opt/kata/
drwxr-xr-x dwg/dwg           0 2022-02-24 16:09 ./opt/kata/bin/
-rwxr--r-- dwg/dwg     3290472 2022-02-24 16:12 ./opt/kata/bin/firecracker
-rwxr--r-- dwg/dwg     2589888 2022-02-24 16:12 ./opt/kata/bin/jailer
-rwxr-xr-x dwg/dwg    18181752 2022-02-24 13:37 ./opt/kata/bin/qemu-system-x86_64
-rwxr-xr-x dwg/dwg       16694 2022-02-24 16:09 ./opt/kata/bin/kata-collect-data.sh
-rwxr-xr-x dwg/dwg    50818680 2022-02-24 16:09 ./opt/kata/bin/kata-runtime
-rwxr-xr-x dwg/dwg    49086688 2022-02-24 16:09 ./opt/kata/bin/containerd-shim-kata-v2
-rwxr-xr-x dwg/dwg    27474130 2022-02-24 16:09 ./opt/kata/bin/kata-monitor
drwxr-xr-x dwg/dwg           0 2022-02-24 16:09 ./opt/kata/share/
drwxr-xr-x dwg/dwg           0 2022-02-24 16:08 ./opt/kata/share/kata-containers/
-rw-r--r-- dwg/dwg     5796400 2022-02-24 16:15 ./opt/kata/share/kata-containers/vmlinuz-5.15.23-89
-rw-r--r-- dwg/dwg    38722544 2022-02-24 16:15 ./opt/kata/share/kata-containers/vmlinux-5.15.23-89
-rw-r--r-- dwg/dwg       72479 2022-02-24 16:15 ./opt/kata/share/kata-containers/config-5.15.23
lrwxrwxrwx dwg/dwg           0 2022-02-24 16:15 ./opt/kata/share/kata-containers/vmlinuz.container -> vmlinuz-5.15.23-89
lrwxrwxrwx dwg/dwg           0 2022-02-24 16:15 ./opt/kata/share/kata-containers/vmlinux.container -> vmlinux-5.15.23-89
drwxr-xr-x dwg/dwg           0 2022-02-24 13:37 ./opt/kata/share/kata-qemu/
drwxr-xr-x dwg/dwg           0 2022-02-24 13:37 ./opt/kata/share/kata-qemu/qemu/
-rw-r--r-- dwg/dwg    67108864 2022-02-24 13:32 ./opt/kata/share/kata-qemu/qemu/edk2-aarch64-code.fd
-rw-r--r-- dwg/dwg    67108864 2022-02-24 13:32 ./opt/kata/share/kata-qemu/qemu/edk2-arm-code.fd
-rw-r--r-- dwg/dwg    67108864 2022-02-24 13:32 ./opt/kata/share/kata-qemu/qemu/edk2-arm-vars.fd
-rw-r--r-- dwg/dwg     3653632 2022-02-24 13:32 ./opt/kata/share/kata-qemu/qemu/edk2-i386-code.fd
-rw-r--r-- dwg/dwg     3653632 2022-02-24 13:32 ./opt/kata/share/kata-qemu/qemu/edk2-i386-secure-code.fd
-rw-r--r-- dwg/dwg      540672 2022-02-24 13:32 ./opt/kata/share/kata-qemu/qemu/edk2-i386-vars.fd
-rw-r--r-- dwg/dwg     3653632 2022-02-24 13:36 ./opt/kata/share/kata-qemu/qemu/edk2-x86_64-code.fd
-rw-r--r-- dwg/dwg     3653632 2022-02-24 13:36 ./opt/kata/share/kata-qemu/qemu/edk2-x86_64-secure-code.fd
drwxr-xr-x dwg/dwg           0 2022-02-24 13:37 ./opt/kata/share/kata-qemu/qemu/vhost-user/
-rw-r--r-- dwg/dwg         121 2022-02-24 13:32 ./opt/kata/share/kata-qemu/qemu/vhost-user/50-qemu-virtiofsd.json
-rw-r--r-- dwg/dwg      131072 2022-02-24 13:31 ./opt/kata/share/kata-qemu/qemu/bios.bin
-rw-r--r-- dwg/dwg      262144 2022-02-24 13:31 ./opt/kata/share/kata-qemu/qemu/bios-256k.bin
-rw-r--r-- dwg/dwg      131072 2022-02-24 13:31 ./opt/kata/share/kata-qemu/qemu/bios-microvm.bin
-rw-r--r-- dwg/dwg       65536 2022-02-24 13:31 ./opt/kata/share/kata-qemu/qemu/qboot.rom
-rw-r--r-- dwg/dwg      160768 2022-02-24 13:31 ./opt/kata/share/kata-qemu/qemu/efi-virtio.rom
-rw-r--r-- dwg/dwg      154542 2022-02-24 13:31 ./opt/kata/share/kata-qemu/qemu/qemu-nsis.bmp
-rw-r--r-- dwg/dwg        1024 2022-02-24 13:31 ./opt/kata/share/kata-qemu/qemu/linuxboot.bin
-rw-r--r-- dwg/dwg        1536 2022-02-24 13:31 ./opt/kata/share/kata-qemu/qemu/linuxboot_dma.bin
-rw-r--r-- dwg/dwg        9216 2022-02-24 13:31 ./opt/kata/share/kata-qemu/qemu/kvmvapic.bin
-rw-r--r-- dwg/dwg        1536 2022-02-24 13:31 ./opt/kata/share/kata-qemu/qemu/pvh.bin
-rw-r--r-- dwg/dwg       50936 2022-02-24 13:31 ./opt/kata/share/kata-qemu/qemu/s390-ccw.img
-rw-r--r-- dwg/dwg       79688 2022-02-24 13:31 ./opt/kata/share/kata-qemu/qemu/s390-netboot.img
-rw-r--r-- dwg/dwg       42903 2022-02-24 13:31 ./opt/kata/share/kata-qemu/qemu/edk2-licenses.txt
-rw-r--r-- dwg/dwg      785696 2022-02-24 13:31 ./opt/kata/share/kata-qemu/qemu/hppa-firmware.img
drwxr-xr-x dwg/dwg           0 2022-02-24 13:37 ./opt/kata/share/kata-qemu/qemu/firmware/
-rw-r--r-- dwg/dwg         746 2022-02-24 13:32 ./opt/kata/share/kata-qemu/qemu/firmware/50-edk2-i386-secure.json
-rw-r--r-- dwg/dwg         771 2022-02-24 13:32 ./opt/kata/share/kata-qemu/qemu/firmware/50-edk2-x86_64-secure.json
-rw-r--r-- dwg/dwg         652 2022-02-24 13:32 ./opt/kata/share/kata-qemu/qemu/firmware/60-edk2-aarch64.json
-rw-r--r-- dwg/dwg         640 2022-02-24 13:32 ./opt/kata/share/kata-qemu/qemu/firmware/60-edk2-arm.json
-rw-r--r-- dwg/dwg         697 2022-02-24 13:32 ./opt/kata/share/kata-qemu/qemu/firmware/60-edk2-i386.json
-rw-r--r-- dwg/dwg         722 2022-02-24 13:32 ./opt/kata/share/kata-qemu/qemu/firmware/60-edk2-x86_64.json
drwxr-xr-x dwg/dwg           0 2022-02-24 16:09 ./opt/kata/share/bash-completion/
drwxr-xr-x dwg/dwg           0 2022-02-24 16:09 ./opt/kata/share/bash-completion/completions/
-rw-r--r-- dwg/dwg        2265 2022-02-24 16:09 ./opt/kata/share/bash-completion/completions/kata-runtime
drwxr-xr-x dwg/dwg           0 2022-02-24 16:09 ./opt/kata/share/defaults/
drwxr-xr-x dwg/dwg           0 2022-02-24 16:09 ./opt/kata/share/defaults/kata-containers/
-rw-r--r-- dwg/dwg       12031 2022-02-24 16:09 ./opt/kata/share/defaults/kata-containers/configuration-clh.toml
-rw-r--r-- dwg/dwg        9318 2022-02-24 16:09 ./opt/kata/share/defaults/kata-containers/configuration-acrn.toml
-rw-r--r-- dwg/dwg       25035 2022-02-24 16:09 ./opt/kata/share/defaults/kata-containers/configuration-qemu.toml
-rw-r--r-- dwg/dwg       15384 2022-02-24 16:09 ./opt/kata/share/defaults/kata-containers/configuration-fc.toml
lrwxrwxrwx dwg/dwg           0 2022-02-24 16:09 ./opt/kata/share/defaults/kata-containers/configuration.toml -> configuration-qemu.toml
drwxr-xr-x dwg/dwg           0 2022-02-24 13:37 ./opt/kata/libexec/
drwxr-xr-x dwg/dwg           0 2022-02-24 13:37 ./opt/kata/libexec/kata-qemu/
-rwxr-xr-x dwg/dwg     2420840 2022-02-24 13:37 ./opt/kata/libexec/kata-qemu/virtiofsd
```

Looking at the individual built log for the rootfs shows an error, which failed to propagate to the top-level build:
```
INFO: DESTDIR /home/dwg/src/kata-containers/tools/packaging/kata-deploy/local-build/build/rootfs-image/destdir
INFO: Create image
build image
~/src/kata-containers/tools/osbuilder ~/src/kata-containers/tools/packaging/kata-deploy/local-build/build/rootfs-image/builddir
INFO: Build image
INFO: image os: clearlinux
INFO: image os version: latest
Creating rootfs for clearlinux
/home/dwg/src/kata-containers/tools/osbuilder/rootfs-builder/rootfs.sh -o 2.4.0-alpha2-6e73dfda5948e23a7cc6f1597019445267cf27d3 -r /home/dwg/src/kata-containers/tools/packaging/kata-deploy/local-build/build/rootfs-image/builddir/rootfs-image/clearlinux_rootfs clearlinux
INFO: Detecting agent rust version
INFO: Get rust version from /home/dwg/src/kata-containers/tools/osbuilder/rootfs-builder/../../../versions.yaml
Required rust version: 1.58.1
INFO: Detecting musl version
INFO: Get musl version from /home/dwg/src/kata-containers/tools/osbuilder/rootfs-builder/../../../versions.yaml
Required musl version: 1.1.23
~/src/kata-containers/tools/osbuilder/rootfs-builder/clearlinux ~/src/kata-containers/tools/osbuilder
~/src/kata-containers/tools/osbuilder
Sending build context to Docker daemon  6.144kB
Step 1/8 : ARG IMAGE_REGISTRY=docker.io
Step 2/8 : FROM ${IMAGE_REGISTRY}/fedora:30
 ---> 177d5adf0c6c
Step 3/8 : RUN dnf -y update && dnf install -y     autoconf     automake     binutils     chrony     coreutils     curl     curl     gcc     gcc-c++     git     glibc-common     glibc-devel     glibc-headers     glibc-static     glibc-utils     libstdc++-devel     libstdc++-static     m4     make     pkgconfig     sed     systemd     tar     vim     which &&     dnf clean all
 ---> Using cache
 ---> eb4b0b8f3943
Step 4/8 : RUN pushd /root;     curl -sLO https://www.musl-libc.org/releases/musl-1.1.23.tar.gz; tar -zxf musl-1.1.23.tar.gz;   cd musl-1.1.23;         sed -i "s/^ARCH = .*/ARCH = x86_64/g" dist/config.mak;  ./configure > /dev/null 2>&1;   make > /dev/null 2>&1;  make install > /dev/null 2>&1;  echo "/usr/local/musl/lib" > /etc/ld-musl-x86_64.path;  popd
 ---> Using cache
 ---> 845e6e7310e7
Step 5/8 : ENV PATH=$PATH:/usr/local/musl/bin
 ---> Using cache
 ---> 98831ad1e039
Step 6/8 : RUN curl --proto '=https' --tlsv1.2 https://sh.rustup.rs -sSLf --output /tmp/rust-init;     chmod a+x /tmp/rust-init;        export http_proxy=;     export https_proxy=;    /tmp/rust-init -y --default-toolchain 1.58.1
 ---> Using cache
 ---> 43210368e0e5
Step 7/8 : RUN . /root/.cargo/env;     export http_proxy=;      export https_proxy=;    cargo install cargo-when;       rustup target install x86_64-unknown-linux-musl
 ---> Using cache
 ---> 4af69e37695b
Step 8/8 : RUN ln -sf /usr/bin/g++ /bin/musl-g++
 ---> Using cache
 ---> 9ab0f2996728
Successfully built 9ab0f2996728
Successfully tagged clearlinux-rootfs-osbuilder:latest
INFO: Detecting agent rust version
INFO: Get rust version from /kata-containers/tools/osbuilder/rootfs-builder/../../../versions.yaml
/home/dwg/go/bin/yq: /lib64/libc.so.6: version `GLIBC_2.32' not found (required by /home/dwg/go/bin/yq)
/home/dwg/go/bin/yq: /lib64/libc.so.6: version `GLIBC_2.34' not found (required by /home/dwg/go/bin/yq)
/home/dwg/go/bin/yq: /lib64/libc.so.6: version `GLIBC_2.32' not found (required by /home/dwg/go/bin/yq)
/home/dwg/go/bin/yq: /lib64/libc.so.6: version `GLIBC_2.34' not found (required by /home/dwg/go/bin/yq)
ERROR: Could not detect the required rust version for AGENT_VERSION='main'.
make: *** [Makefile:93: /home/dwg/src/kata-containers/tools/packaging/kata-deploy/local-build/build/rootfs-image/builddir/rootfs-image/.clearlinux_rootfs.done] Error 1
./
./opt/
./opt/kata/
./opt/kata/share/
./opt/kata/share/kata-containers/
drwxr-xr-x dwg/dwg           0 2022-02-24 05:08 ./
drwxr-xr-x dwg/dwg           0 2022-02-24 05:08 ./opt/
drwxr-xr-x dwg/dwg           0 2022-02-24 05:08 ./opt/kata/
drwxr-xr-x dwg/dwg           0 2022-02-24 05:08 ./opt/kata/share/
drwxr-xr-x dwg/dwg           0 2022-02-24 05:08 ./opt/kata/share/kata-containers/
```

# Further information

I suspect this is a mismatch between libc versions in the host and docker build environments, combined with sloppy handling of whether we're invoking things from mounted host directories or internally built copies.

I discovered the cause of the error here, but immediately hit another one.

The GLIBC version error was caused because I somehow had an existing, dynamically linked `~/go/bin/yq` on my host system.  The build was attempting to use this and failing because it didn't match the libc version within the build docker container.  This is still a bug: if we're using a docker build, we shouldn't be relying on things from the host that are outside the source tree we're building.

In addition, when I worked around this by deleting the dynamic binary, the build *still* silently failed, now with this error in one of the sub logfiles.  As before the error was not properly propagated to the top level.

```
 ---> f1f97e2ec652
Successfully built f1f97e2ec652
Successfully tagged alpine-rootfs-osbuilder:latest
INFO: rootfs_lib.sh file found. Loading content
INFO: Detecting agent rust version
INFO: Get rust version from /kata-containers/tools/osbuilder/rootfs-builder/../../../versions.yaml
Required rust version: 1.58.1
INFO: Detecting musl version
INFO: Get musl version from /kata-containers/tools/osbuilder/rootfs-builder/../../../versions.yaml
Required musl version: 1.1.23
INFO: build directly
fetch https://dl-5.alpinelinux.org/alpine/v3.15/main/x86_64/APKINDEX.tar.gz
(1/21) Installing musl (1.2.2-r7)
(2/21) Installing busybox (1.34.1-r4)
Executing busybox-1.34.1-r4.post-install
(3/21) Installing alpine-baselayout (3.2.0-r18)
Executing alpine-baselayout-3.2.0-r18.pre-install
Executing alpine-baselayout-3.2.0-r18.post-install
(4/21) Installing ifupdown-ng (0.11.3-r0)
(5/21) Installing openrc (0.44.7-r3)
Executing openrc-0.44.7-r3.post-install
(6/21) Installing alpine-conf (3.13.0-r1)
(7/21) Installing ca-certificates-bundle (20211220-r0)
(8/21) Installing libcrypto1.1 (1.1.1l-r8)
(9/21) Installing libssl1.1 (1.1.1l-r8)
(10/21) Installing libretls (3.3.4-r2)
(11/21) Installing ssl_client (1.34.1-r4)
(12/21) Installing zlib (1.2.11-r3)
(13/21) Installing apk-tools (2.12.7-r3)
(14/21) Installing busybox-suid (1.34.1-r4)
(15/21) Installing busybox-initscripts (4.0-r5)
Executing busybox-initscripts-4.0-r5.post-install
(16/21) Installing scanelf (1.3.3-r0)
(17/21) Installing musl-utils (1.2.2-r7)
(18/21) Installing libc-utils (0.7.2-r3)
(19/21) Installing alpine-keys (2.4-r1)
(20/21) Installing alpine-base (3.15.0-r0)
(21/21) Installing libseccomp (2.5.2-r0)
Executing busybox-1.34.1-r4.trigger
OK: 9 MiB in 21 packages
INFO: Create symlink to /tmp in /var to create private temporal directories with systemd
INFO: Create /rootfs/etc
INFO: Configure chrony file /rootfs/etc/chrony.conf
/kata-containers/tools/osbuilder/rootfs-builder/rootfs.sh: line 558: rustup: command not found
fatal: could not create work tree dir '/home/dwg/go/src/github.com/kata-containers/tests': File exists
Failed at 563: bash ${script_dir}/../../../ci/install_rust.sh ${RUST_VERSION}
make: *** [Makefile:93: /home/dwg/src/kata-containers/tools/packaging/kata-deploy/local-build/build/rootfs-initrd/builddir/initrd-image/.alpine_rootfs.done] Error 128
```

--- a/tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh
+++ b/tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh
@@ -64,6 +64,7 @@ version: The kata version that will be use to create the tarball
 options:
 
 -h|--help      	      : Show this help
+-s             	      : Silent mode (produce output in case of failure only)
 --build=<asset>       :
 	all
 	cloud-hypervisor

--- a/tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh
+++ b/tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh
@@ -8,6 +8,7 @@
 set -o errexit
 set -o nounset
 set -o pipefail
+set -o errtrace
 
 readonly project="kata-containers"
 
@@ -196,6 +197,18 @@ handle_build() {
 	tar tvf "${tarball_name}"
 }
 
+silent_mode_error_trap() {
+	local stdout="$1"
+	local stderr="$2"
+	local t="$3"
+	local log_file="$4"
+	exec 1>&${stdout}
+	exec 2>&${stderr}
+	error "Failed to build: $t, logs:"
+	cat "${log_file}"
+	exit 1
+}
+
 main() {
 	local build_targets
 	local silent
@@ -248,11 +261,15 @@ main() {
 		(
 			cd "${builddir}"
 			if [ "${silent}" == true ]; then
-				if ! handle_build "${t}" &>"$log_file"; then
-					error "Failed to build: $t, logs:"
-					cat "${log_file}"
-					exit 1
-				fi
+				local stdout
+				local stderr
+				# Save stdout and stderr, to be restored
+				# by silent_mode_error_trap() in case of
+				# build failure.
+				exec {stdout}>&1
+				exec {stderr}>&2
+				trap "silent_mode_error_trap $stdout $stderr $t \"$log_file\"" ERR
+				handle_build "${t}" &>"$log_file"
 			else
 				handle_build "${t}"
 			fi

